File: src/models/media.rs
use serde::{Deserialize, Serialize};
use sqlx::prelude::Type;
use strum::{AsRefStr, Display, EnumString};
use uuid::Uuid;

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
//  Enums
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
    Default,
    Type,
    Display,
    EnumString,
    AsRefStr,
)]
#[sqlx(type_name = "TEXT", rename_all = "UPPERCASE")]
pub enum MediaProvider {
    #[default]
    MAL,
    ANILIST,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub enum MediaType {
    #[default]
    Anime,
    Manga,
}

/// What format the media was released as.
///
/// Unified superset of MAL media_type + AniList MediaFormat.
/// MAL anime: unknown, tv, ova, movie, special, ona, music
/// MAL manga: unknown, manga, novel, one_shot, doujinshi, manhwa, manhua, oel
/// AniList:   TV, TV_SHORT, MOVIE, SPECIAL, OVA, ONA, MUSIC, MANGA, NOVEL, ONE_SHOT
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub enum MediaFormat {
    #[default]
    Unknown,
    Tv,
    TvShort,
    Movie,
    Special,
    Ova,
    Ona,
    Music,
    Manga,
    Novel,
    OneShot,
    Doujinshi,
    Manhwa,
    Manhua,
    Oel,
}

/// Airing / publishing status of the media itself.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub enum ReleaseStatus {
    #[default]
    Unknown,
    Releasing,
    Finished,
    NotYetReleased,
    Cancelled,
    Hiatus,
}

/// User's personal watching/reading status.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub enum ListStatus {
    #[default]
    Current,
    Planning,
    Completed,
    Dropped,
    Paused,
    Repeating,
}

/// NSFW classification.
/// MAL distinguishes white/gray/black. AniList only has isAdult bool.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub enum NsfwLevel {
    #[default]
    Safe,
    Gray,
    Nsfw,
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
//  Value objects
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct MediaTitle {
    /// AniList: romaji. MAL: title (the main title IS romanized).
    pub romanized: Option<String>,
    /// AniList: english. MAL: alternative_titles.en.
    pub english: Option<String>,
    /// AniList: native. MAL: alternative_titles.ja.
    pub native: Option<String>,
    /// AniList: userPreferred. MAL: same as romanized.
    pub user_preferred: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct CoverImage {
    /// AniList only. Falls back to large if unavailable.
    pub extra_large: Option<String>,
    /// Both APIs.
    pub large: Option<String>,
    /// Both APIs.
    pub medium: Option<String>,
    /// AniList only. Average hex color of the cover.
    pub color: Option<String>,
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
//  Core models
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

/// The media itself — anime or manga metadata.
/// No user state. No list status. Just what the thing IS.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct Media {
    pub id: Uuid,

    /// Provider-specific ID (MAL int or AniList int, as string).
    pub provider_id: String,
    pub provider: MediaProvider,

    pub media_type: MediaType,
    pub format: MediaFormat,
    pub release_status: ReleaseStatus,

    pub title: MediaTitle,
    pub cover: CoverImage,
    pub synopsis: Option<String>,

    /// Community mean score. MAL: 0.0–10.0. AniList: 0–100 (normalize to 0–10).
    pub mean_score: Option<f32>,
    pub popularity: Option<i32>,

    // ── Type-specific counts ──
    /// Anime only. Total episode count (0 or null if unknown).
    pub episodes: Option<i32>,
    /// Anime only. Episode length in seconds.
    pub duration: Option<i32>,
    /// Manga only.
    pub chapters: Option<i32>,
    /// Manga only.
    pub volumes: Option<i32>,

    pub genres: Vec<String>,
    pub nsfw: NsfwLevel,
}

/// What the user thinks of a specific media — their list entry.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct ListEntry {
    pub status: ListStatus,

    /// User's score. Normalized to 0.0–10.0.
    pub score: Option<f32>,

    /// Episodes watched (anime) or chapters read (manga).
    pub progress: Option<i32>,
    /// Volumes read (manga only).
    pub progress_volumes: Option<i32>,

    pub is_repeating: bool,
    pub repeat_count: Option<i32>,

    pub tags: Vec<String>,
    /// AniList: notes. MAL: comments.
    pub notes: Option<String>,

    pub updated_at: Option<String>,
}

/// The "edge" that joins a media with a user's list entry.
/// This is what the API returns: one of these per item in the list.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct MediaEntry {
    pub media: Media,
    pub list_entry: ListEntry,
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
//  Pagination
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct Paging {
    /// MAL provides cursor URLs. AniList uses page numbers.
    /// Store the raw cursor/URL if the provider gives one.
    pub next_cursor: Option<String>,
    pub prev_cursor: Option<String>,
    pub has_next: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct PaginatedResponse {
    pub data: Vec<MediaEntry>,
    pub paging: Paging,
}

File: src/dto/media.go
package dto

import "uuid"

// MediaFormat defines what format the media was released as.
// Unified superset of MAL media_type + AniList MediaFormat.
// MAL anime: unknown, tv, ova, movie, special, ona, music
// MAL manga: unknown, manga, novel, one_shot, doujinshi, manhwa, manhua, oel
// AniList:   TV, TV_SHORT, MOVIE, SPECIAL, OVA, ONA, MUSIC, MANGA, NOVEL, ONE_SHOT
type MediaFormat string

const (
	MediaFormatUnknown   MediaFormat = "Unknown"
	MediaFormatTv        MediaFormat = "Tv"
	MediaFormatTvShort   MediaFormat = "TvShort"
	MediaFormatMovie     MediaFormat = "Movie"
	MediaFormatSpecial   MediaFormat = "Special"
	MediaFormatOva       MediaFormat = "Ova"
	MediaFormatOna       MediaFormat = "Ona"
	MediaFormatMusic     MediaFormat = "Music"
	MediaFormatManga     MediaFormat = "Manga"
	MediaFormatNovel     MediaFormat = "Novel"
	MediaFormatOneShot   MediaFormat = "OneShot"
	MediaFormatDoujinshi MediaFormat = "Doujinshi"
	MediaFormatManhwa    MediaFormat = "Manhwa"
	MediaFormatManhua    MediaFormat = "Manhua"
	MediaFormatOel       MediaFormat = "Oel"
)

// ReleaseStatus defines the airing / publishing status of the media itself.
type ReleaseStatus string

const (
	ReleaseStatusUnknown        ReleaseStatus = "Unknown"
	ReleaseStatusReleasing      ReleaseStatus = "Releasing"
	ReleaseStatusFinished       ReleaseStatus = "Finished"
	ReleaseStatusNotYetReleased ReleaseStatus = "NotYetReleased"
	ReleaseStatusCancelled      ReleaseStatus = "Cancelled"
	ReleaseStatusHiatus         ReleaseStatus = "Hiatus"
)

// ListStatus defines user's personal watching/reading status.
type ListStatus string

const (
	ListStatusCurrent   ListStatus = "Current"
	ListStatusPlanning  ListStatus = "Planning"
	ListStatusCompleted ListStatus = "Completed"
	ListStatusDropped   ListStatus = "Dropped"
	ListStatusPaused    ListStatus = "Paused"
	ListStatusRepeating ListStatus = "Repeating"
)

// NsfwLevel defines NSFW classification.
// MAL distinguishes white/gray/black. AniList only has isAdult bool.
type NsfwLevel string

const (
	NsfwLevelSafe NsfwLevel = "Safe"
	NsfwLevelGray NsfwLevel = "Gray"
	NsfwLevelNsfw NsfwLevel = "Nsfw"
)

type MediaProvider string

const (
	MediaProviderMAL     MediaProvider = "MAL"
	MediaProviderAnilist MediaProvider = "ANILIST"
)

type MediaType string

const (
	MediaTypeAnime MediaType = "Anime"
	MediaTypeManga MediaType = "Manga"
	MediaTypeNovel MediaType = "Novel"
)

type MediaTitle struct {
	// AniList: romaji. MAL: title (the main title IS romanized).
	Romanized *string `json:"romanized"`
	// AniList: english. MAL: alternative_titles.en.
	English *string `json:"english"`
	// AniList: native. MAL: alternative_titles.ja.
	Native *string `json:"native"`
	// AniList: userPreferred. MAL: same as romanized.
	UserPreferred *string `json:"user_preferred"`
}

type CoverImage struct {
	// AniList only. Falls back to large if unavailable.
	ExtraLarge *string `json:"extra_large"`
	// Both APIs.
	Large *string `json:"large"`
	// Both APIs.
	Medium *string `json:"medium"`
	// AniList only. Average hex color of the cover.
	Color *string `json:"color"`
}

// Media represents the media itself — anime or manga metadata.
// No user state. No list status. Just what the thing IS.
type Media struct {
	ID uuid.UUID `json:"id"`

	// Provider-specific ID (MAL int or AniList int, as string).
	ProviderID string        `json:"provider_id"`
	Provider   MediaProvider `json:"provider"`

	MediaType     MediaType     `json:"media_type"`
	Format        MediaFormat   `json:"format"`
	ReleaseStatus ReleaseStatus `json:"release_status"`

	Title    MediaTitle `json:"title"`
	Cover    CoverImage `json:"cover"`
	Synopsis *string    `json:"synopsis"`

	// Community mean score. MAL: 0.0–10.0. AniList: 0–100 (normalize to 0–10).
	MeanScore  *float32 `json:"mean_score"`
	Popularity *int     `json:"popularity"`

	// Type-specific counts
	// Anime only. Total episode count (0 or null if unknown).
	Episodes *int `json:"episodes"`
	// Anime only. Episode length in seconds.
	Duration *int `json:"duration"`
	// Manga only.
	Chapters *int `json:"chapters"`
	// Manga only.
	Volumes *int `json:"volumes"`

	Genres []string  `json:"genres"`
	Nsfw   NsfwLevel `json:"nsfw"`
}

// ListEntry represents what the user thinks of a specific media — their list entry.
type ListEntry struct {
	Status ListStatus `json:"status"`

	// User's score. Normalized to 0.0–10.0.
	Score *float32 `json:"score"`

	// Episodes watched (anime) or chapters read (manga).
	Progress *int `json:"progress"`
	// Volumes read (manga only).
	ProgressVolumes *int `json:"progress_volumes"`

	IsRepeating bool `json:"is_repeating"`
	RepeatCount *int `json:"repeat_count"`

	Tags []string `json:"tags"`
	// AniList: notes. MAL: comments.
	Notes *string `json:"notes"`

	UpdatedAt *string `json:"updated_at"`
}

// MediaEntry is the "edge" that joins a media with a user's list entry.
// This is what the API returns: one of these per item in the list.
type MediaEntry struct {
	Media     Media     `json:"media"`
	ListEntry ListEntry `json:"list_entry"`
}

type Paging struct {
	// MAL provides cursor URLs. AniList uses page numbers.
	// Store the raw cursor/URL if the provider gives one.
	NextCursor *string `json:"next_cursor"`
	PrevCursor *string `json:"prev_cursor"`
	HasNext    bool    `json:"has_next"`
}

type PaginatedResponse struct {
	Data   []MediaEntry `json:"data"`
	Paging Paging       `json:"paging"`
}
